### PR TITLE
Add a banded sliding-window SDPA fast path for Gemma-4 style layers

### PR DESCRIPTION
**Pull request by danielhanchen**
## Summary

Sliding-window attention only needs keys in `(i-w, i]` for query i, but SDPA with an explicit band mask still materializes the full O(seq^2) score matrix. On gemma-4 the 25 sliding layers dominate step time at 16k/32k tokens whenever flash-attn is unavailable (the head_dim 512 global layers keep FA2 off model-wide).

## What this does

Adds a banded (block-local) O(seq*w) formulation on stock SDPA: cut the sequence into w-sized blocks, let query block b attend its own and the previous key block (length 2w) under a single static (w, 2w) causal+locality mask reused for every block and layer (block 0 masks its phantom previous block), and fold blocks into the batch dimension for one SDPA call. Backward is exact through SDPA.

Opt-in via `UNSLOTH_BANDED_SDPA=1`. It engages from the sdpa wrapper only for batch size 1 and sliding layers with head_dim at most 256 whose mask is verified to be the plain causal+sliding band with no token padding (cheap probe rows, cached per mask); anything else defers to the original SDPA. Needs no flash-attn build and works with any SDPA backend and dtype, so it complements the FA2 sliding path as the portable fallback.

## Testing

- 6 CPU unit tests: parity vs full SDPA + band mask in fp32 (rel 1e-5) across GQA/MQA and ragged tail blocks, gradient parity, block-mask structure, and band-mask detection.
- Real-model layer-0 attention output rel 1.2e-5 vs full SDPA on 4-bit weights.
- Measured on a 26B-A4B 4-bit LoRA SFT: 1.21x at 8k, 2.48x at 16k, 2.87x at 32k with unchanged peak memory (independently reproduced: 3738 vs 1531 tok/s at 16k).

### CODE REVIEW

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
Same mask fixes as the flash-sliding path: the verdict is cached on the mask tensor rather than an `id()`-keyed dict, and the band check verifies the full mask so a packed mask defers to SDPA instead of being sampled. Also changed `.view` to `.reshape` on the query for non-contiguous inputs. Added tests for the id-reuse and non-probe-row cases.

File: unsloth_zoo/temporary_patches/gemma4_banded_attention.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    if m.shape[-2] != S or m.shape[-1] != S:
+        ok = False
+    else:
+        allowed = m if m.dtype == torch.bool else (m > -1e4)
+        idx = torch.arange(S, device=m.device)
+        band = (idx[None, :] <= idx[:, None]) & (idx[None, :] > idx[:, None] - w)
+        ok = bool((allowed == band).all().item())
+    try:
+        mask._unsloth_plain_band = ok
+    except Exception:
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_banded_attention.py:110]: The bare except is intentional. The plain-band verdict is a best-effort cache write onto the mask tensor; a few tensor types reject custom attributes, and a failed write is non-fatal since the next call just recomputes. This matches the existing except: pass usage across temporary_patches.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/gemma4_banded_attention.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    if cached is not None:
+        return cached
+    m = mask[0, 0]
+    if m.shape[-2] != S or m.shape[-1] != S:
+        ok = False
+    else:
+        allowed = m if m.dtype == torch.bool else (m > -1e4)
+        idx = torch.arange(S, device=m.device)
+        band = (idx[None, :] <= idx[:, None]) & (idx[None, :] > idx[:, None] - w)
+        ok = bool((allowed == band).all().item())
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_banded_attention.py]: Applied the same row-chunked verification as the flash-sliding path, so we no longer rebuild a dense S x S band plus its comparison at long context.

File: unsloth_zoo/temporary_patches/gemma4_banded_attention.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    object id.
+    """
+    if mask is None:
+        return True
+    if not torch.is_tensor(mask) or mask.dim() != 4:
+        return False
+    cached = getattr(mask, "_unsloth_plain_band", None)
+    if cached is not None:
+        return cached
+    m = mask[0, 0]
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_banded_attention.py]: Added a head guard: masks with shape[1] > 1 are now rejected outright, since the single block mask is applied to every head and a per-head mask cannot be honored.

File: unsloth_zoo/temporary_patches/gemma4_banded_attention.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    if m.shape[-2] != S or m.shape[-1] != S:
+        ok = False
+    else:
+        allowed = m if m.dtype == torch.bool else (m > -1e4)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_banded_attention.py]: Tightened the float check: in-band entries must be exactly 0 and out-of-band entries <= -1e4, otherwise we fall back to normal SDPA. Applied the same strictness to the flash-sliding path so both behave identically.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/gemma4_banded_attention.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        _BAND_OK.clear()
+    m = mask[0, 0]
+    if m.shape[-2] != S or m.shape[-1] != S:
+        _BAND_OK[key] = False
+        return False
+    allowed = m if m.dtype == torch.bool else (m > -1e4)
+    dev = m.device
+    cand = sorted({x for x in (0, 1, w - 1, w, w + 1, 2 * w, S // 2, S - 1, S - w, S - w - 1) if 0 <= x < S})
+    probes = torch.tensor(cand, device=dev)
+    rows = allowed[probes]                                # (P, S)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_banded_attention.py]: The verifier now scans every row in chunks, so a packed or padded mask with an interior boundary is rejected and we fall back to full SDPA. Covered by test_mask_is_plain_band_rejects_non_probe_violation.

File: unsloth_zoo/temporary_patches/gemma4_banded_attention.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+import os
+import torch
+import torch.nn.functional as F
+from .common import logger
+
+__all__ = ["maybe_banded_sliding"]
+
+_MASK_CACHE = {}    # (w, nb, device) -> (nb,1,w,2w) bool block mask
+_BAND_OK = {}       # id(attention_mask) -> bool (is it a plain no-padding band)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_banded_attention.py]: Dropped the id-keyed dict. The band verdict is cached on the mask tensor via _unsloth_plain_band, so a recycled object id can no longer return a stale result. Covered by test_mask_is_plain_band_survives_id_reuse.

File: unsloth_zoo/temporary_patches/gemma4_banded_attention.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    if num_key_value_groups > 1:
+        key = key[:, :, None].expand(B, Hkv, num_key_value_groups, S, d).reshape(B, H, S, d)
+        value = value[:, :, None].expand(B, Hkv, num_key_value_groups, S, d).reshape(B, H, S, d)
+    Sp = ((S + w - 1) // w) * w
+    if Sp != S:
+        query = F.pad(query, (0, 0, 0, Sp - S))
+        key = F.pad(key, (0, 0, 0, Sp - S))
+        value = F.pad(value, (0, 0, 0, Sp - S))
+    nb = Sp // w
+    qb = query.view(B, H, nb, w, d).permute(0, 2, 1, 3, 4).reshape(B * nb, H, w, d)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_banded_attention.py]: Switched both the query fold and the output reshape to .reshape, so a non-contiguous query no longer raises.

**Comment by danielhanchen** [comment]
### Final results

Setup: banded (block-local) sliding-window SDPA fast path (`UNSLOTH_BANDED_SDPA=1`), measured on a 26B-A4B 4-bit LoRA SFT, peak memory unchanged.

| seq len | speedup | tok/s (before -> after) |
|---|---|---|
| 8k | 1.21x | n/a |
| 16k | 2.48x | 1531 -> 3738 |
| 32k | 2.87x | n/a |

The 16k tok/s pair was independently reproduced. Layer-0 attention output matches full SDPA to rel 1.2e-5 on 4-bit weights.

Takeaway: a portable SDPA fallback for sliding-window layers that needs no flash-attn build, up to 2.87x faster at 32k tokens with unchanged peak memory.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/gemma4_banded_attention.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+# cut the sequence into w-sized blocks, let query block b attend key blocks b-1
+# and b (length 2w) under one static (w, 2w) mask (block 0 masks its phantom
+# previous block), and fold blocks into the batch dim for a single SDPA call.
+# Backward is exact through SDPA; fp32 rel ~1e-8 vs full SDPA + band mask.
+# Engaged only when the mask is exactly the causal+sliding band with no token
+# padding (fully verified, so packed masks defer), else defers to SDPA.
+# Opt-in UNSLOTH_BANDED_SDPA=1.
+# ============================================================================
+
+import os
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_banded_attention.py:31]: Good catch, this was a real CI failure. Fixed in 20ba6158: added unsloth_zoo.temporary_patches.gemma4_banded_attention to TEMPORARY_PATCHES_SUBMODULES so the import-smoke completeness test passes (the module is a real helper exporting maybe_banded_sliding, imported by misc.py). test_temporary_patches_imports.py is green (30 passed).

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
Superseded by #860, which merged the portable banded sliding-window SDPA fast path (gemma4_banded_attention.py) into the unified default-on gemma-4 sliding-window router alongside the FlashAttention-2 path. #860 carries this file and its tests verbatim, so the banded work ships in main via #860. Closing to avoid duplicate/overlapping changes.
